Source/Core/Common/FileUtil.cpp

```diff
@@ -298,130 +268,68 @@ bool RenameSync(const std::string& srcFilename, const std::string& destFilename)
   return true;
 }
 
-// copies file srcFilename to destFilename, returns true on success
-bool Copy(const std::string& srcFilename, const std::string& destFilename)
+// copies file srcPath to destPath, returns true on success
+bool Copy(const Path& src_path, const std::string& dest_path)
 {
-  INFO_LOG(COMMON, "Copy: %s --> %s", srcFilename.c_str(), destFilename.c_str());
-#ifdef _WIN32
-  if (CopyFile(UTF8ToTStr(srcFilename).c_str(), UTF8ToTStr(destFilename).c_str(), FALSE))
-    return true;
+  INFO_LOG(COMMON, "Copy: %s --> %s", src_path.GetName().c_str(), dest_path.c_str());
 
-  ERROR_LOG(COMMON, "Copy: failed %s --> %s: %s", srcFilename.c_str(), destFilename.c_str(),
-            GetLastErrorMsg().c_str());
-  return false;
-#else
-
-// buffer size
-#define BSIZE 1024
-
-  char buffer[BSIZE];
+  static constexpr size_t BUFFER_SIZE = 1024;
+  char buffer[BUFFER_SIZE];
 
   // Open input file
-  std::ifstream input;
-  OpenFStream(input, srcFilename, std::ifstream::in | std::ifstream::binary);
-  if (!input.is_open())
+  std::unique_ptr<ReadOnlyFile> input_file = src_path.OpenFile(true);
+  ReadOnlyFileStream input(input_file.get());
+  if (input.fail())
   {
-    ERROR_LOG(COMMON, "Copy: input failed %s --> %s: %s", srcFilename.c_str(), destFilename.c_str(),
-              GetLastErrorMsg().c_str());
+    ERROR_LOG(COMMON, "Copy: input failed %s --> %s: %s", src_path.GetName().c_str(),
+              dest_path.c_str(), GetLastErrorMsg().c_str());
     return false;
   }
 
   // open output file
-  File::IOFile output(destFilename, "wb");
+  File::IOFile output(dest_path, "wb");
 
   if (!output.IsOpen())
   {
-    ERROR_LOG(COMMON, "Copy: output failed %s --> %s: %s", srcFilename.c_str(),
-              destFilename.c_str(), GetLastErrorMsg().c_str());
+    ERROR_LOG(COMMON, "Copy: output failed %s --> %s: %s", src_path.GetName().c_str(),
+              dest_path.c_str(), GetLastErrorMsg().c_str());
     return false;
   }
 
   // copy loop
   while (!input.eof())
   {
     // read input
-    input.read(buffer, BSIZE);
+    input.read(buffer, BUFFER_SIZE);
     if (!input)
     {
-      ERROR_LOG(COMMON, "Copy: failed reading from source, %s --> %s: %s", srcFilename.c_str(),
-                destFilename.c_str(), GetLastErrorMsg().c_str());
+      ERROR_LOG(COMMON, "Copy: failed reading from source, %s --> %s: %s",
+                src_path.GetName().c_str(), dest_path.c_str(), GetLastErrorMsg().c_str());
       return false;
     }
 
     // write output
-    if (!output.WriteBytes(buffer, BSIZE))
+    if (!output.WriteBytes(buffer, BUFFER_SIZE))
```

Comment:
How does this work properly for files whose sizes aren't a multiple of the buffer size? Our file copying implementation hasn't been broken all along, right?